vowpalwabbit/global_data.cc

```diff
@@ -272,6 +300,86 @@ vw_ostream::vw_ostream() : std::ostream(&buf), buf(*this), trace_context(nullptr
 VW_WARNING_STATE_PUSH
 VW_WARNING_DISABLE_DEPRECATED_USAGE
 
+void vw::copy_prediction(const polyprediction& from_pred)
+{
+  if(l == nullptr)
+    return;
+
+  switch(l->pred_type)
+  {
+    case prediction_type_t::action_probs:
+      copy_array(_predict_buffer.a_s, from_pred.a_s);
+      break;
+    case prediction_type_t::action_scores:
+      copy_array(_predict_buffer.a_s, from_pred.a_s);
+      break;
+    case prediction_type_t::decision_probs:
+      copy_array(_predict_buffer.decision_scores, from_pred.decision_scores);
+      break;
+    case prediction_type_t::multilabels:
+      copy_array(_predict_buffer.multilabels.label_v, from_pred.multilabels.label_v);
+      break;
+    case prediction_type_t::pdf:
+      copy_array(_predict_buffer.pdf, from_pred.pdf);
+      break;
+    case prediction_type_t::scalars:
+      copy_array(_predict_buffer.scalars, from_pred.scalars);
+      break;
+    case prediction_type_t::multiclass:
+      _predict_buffer = from_pred;
+      break;
+    case prediction_type_t::action_pdf_value:
+      _predict_buffer = from_pred;
+      break;
+    case prediction_type_t::prob:
+      _predict_buffer = from_pred;
+      break;
+    case prediction_type_t::scalar:
+      _predict_buffer = from_pred;
+      break;
+    default:
+        THROW("Unhandled prediction type");
+  }
+}
+
+void vw::cleanup_prediction()
```

Comment:
This whole function just hammers home how desperately we need these types to be able to have destructors

Reply:
Yeah.  But unions...